### ISSUE TITLE
feature request: update the database schema to support filtering

### ISSUE THREAD

**Issue by mrparkers** [grafeas/grafeas#457]
**Is your feature request related to a problem? Please describe.**

We are looking to add filtering support for PostgreSQL. We already have a proof of concept for the querying based on the work completed by @clrprod, [here](https://github.com/liatrio/grafeas/blob/filterParser/go/v1beta1/storage/pgsqlstore.go). However, we are finding it quite difficult to complete this work due to the existing database schema. 

For example, the occurrence table only contains an `id`, a `project_name`, an `occurrence name`, and a large unstructured `data` column. This makes it challenging to filter for things like the resource uri using SQL. This column can’t be compared using a basic WHERE clause, and the high cardinality of the column means that it can’t be properly indexed by the database.

**Describe the solution you'd like**

We are proposing a change to the schema of Grafea’s storage backend, of the fields from `data`. This could be something smaller, like placing a few fields into new columns e.g. `resource_name`, `resource_uri`, `kind`, `createTime`, and `updateTime`. Or it could be something larger, like representing all of the different types of occurrences in the data layer itself.

To prevent this from becoming a breaking change, we are thinking that database migration functionality may need to be added to the project as well, to handle moving existing databases from the old schema to the new one. We wrote a small proof of concept to demonstrate what this might look like.  You can view the changes [here](https://github.com/grafeas/grafeas/compare/master...liatrio:postgres-schema-poc)

This schema migration example is only meant to communicate the basic idea, to see if this kind of solution would be acceptable going forward.  This schema upgrade would allow us to filter on fields such as `kind` and `resource_uri`, but would not allow us to (for example) select all vulnerability occurrences with a severity of “critical”. Filtering on this type of information would require additional tables to represent each type of occurrence using columns that can be compared using SQL WHERE clauses. We can discuss this further if you’d like.

**Describe alternatives you've considered**

Using the schema as-is, and perform the `data` filtering in memory of our own application. This would require loading essentially all of the records for a given resource into this consumer application, or fronting Grafeas with our own caching solution.
Adding a different storage backend that could better handle filtering of the unstructured or document-style `data` column. 
Reimplementing Grafeas resource concepts over our own storage solution, no longer using Grafeas directly.

**Additional context**

The filtering functionality shared uses the already existing filter code with cel. This does not seem to be currently used or wired up. We aren’t sure about the full context around this existing code, so feel free to provide some feedback about what we could be doing better here.

Our current PoC solution is in upstream, not the separate [pgsql repository](https://github.com/grafeas/grafeas-pgsql). It is not clear when that will cutover, and if this work should go there.

The schema change we’re suggesting would be incompatible with existing versions of Grafeas that already have notes and occurrences within the database. We should discuss the best way to handle a change like this. The schema migration framework we selected has the ability to run migrations in steps, so we could use this to run a step that adds the required columns, then use golang to update each occurrence row to utilize these new columns, then run a step that adds a `NOT NULL` constraint to these columns. Please let us know what you think is the best route here.

Thanks!

**Comment by aysylu** [comment]
Hi @mrparkers,

Thank you for filing the feature request. This FR sounds like an improvement and I'm sure other community members will benefit from the solution. Regarding changes in grafeas vs [grafeas-pgsql](https://github.com/grafeas/grafeas-pgsql), I recommend implementing them in the latter. @judavi started this migration and this is consistent with the long-term plans to have storage backends separate from the API (#341).

Contributions to complete the migration (and/or confirm its completion) and implement the changes for FR (which it sounds like you already have a draft for) are welcome!

Thanks,
Aysylu

**Comment by jhill-cpu** [comment]
I just wanted to ask if there would be any follow up with this issue? My team is moving forward with Grafeas but we need filtering, so we'll probably fork the repo and do it separately since it seems like this issue has been inactive for quite a while. 

Let us know either way! We'd be happy to try pushing our changes back upstream if possible.

**Comment by mrparkers** [comment]
No, we decided to pivot and write our own backend using Elasticsearch, which you can view here: https://github.com/rode/grafeas-elasticsearch

It isn't a complete implementation, but imo it works better than the default postgres backend due to the filtering support we added and the ability to run version migrations on the grafeas schema.
